### PR TITLE
fix crash if ddl item is nil

### PR DESCRIPTION
**Pull request by megan07**
Fixes https://github.com/hashicorp/terraform-provider-google/issues/12974

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:bug
spanner: fixed crash when `google_spanner_database.ddl` item was nil
```

### CODE REVIEW

**Review by SarahFrench — COMMENTED** [review]
How are you reproducing this issue?

I'm trying to reproduce it with a config where `ddl` is set to `[]` or `null` but I'm not experiencing the crash. Is an old version of TF needed to see the issue?

**Comment by megan07** [comment]
An empty string - it took me awhile to reproduce it, but in the postgres test I just changed it and ran it and could see it.

**Comment by SarahFrench** [comment]
Ah ok I've reproduced the error and confirmed that this fix works.

I did find that if a user ends up setting `ddl = [ "" ]` in their TF config then they'll get an API error:

Screenshot 2022-11-10 at 15 45 35

Is this something we need to protect against as well?

**Review by SarahFrench — COMMENTED** [review]
This PR fixes the issue when creating a new resource, but it looks there needs to be the same fix applied to the update action, at this point:

https://github.com/GoogleCloudPlatform/magic-modules/blob/4f6c2ab3e5ae84dec68802e4f40a0005fa769c96/mmv1/templates/terraform/update_encoder/spanner_database.go.erb#L20-L23

File: mmv1/templates/terraform/update_encoder/spanner_database.go.erb [hunk 1]
```diff
@@ -19,17 +19,19 @@ updateDdls := []string{}
 
 //Only new ddl statments to be add to update call
 for i := len(oldDdls); i < len(newDdls); i++ {
-    updateDdls = append(updateDdls, newDdls[i].(string))
+	if newDdls != nil {
+		updateDdls = append(updateDdls, newDdls[i].(string))
+	}
 }
```
Comment (SarahFrench) [inline mmv1/templates/terraform/update_encoder/spanner_database.go.erb:25]: Could you please update to `if newDdls[i] != nil {` ?

**Review by SarahFrench — COMMENTED** [review]
Thanks for bearing with me on this!

There's a small change needed to help the update code run without crashing, and also I've made suggestions for avoiding API errors once the crashing is stopped (see comments)

BTW, I've tested those changes manually and it's worked as expected:

Creating a resource like this:

```hcl
resource "google_spanner_instance" "main" {
  config       = "regional-${var.default_region}"
  display_name = "main-instance"
  num_nodes    = 1
}

resource "google_spanner_database" "database" {
  instance = google_spanner_instance.main.name
  name     = "my-database"
  ddl = [
    "",
  ]
  deletion_protection = false
}
```

and then triggering an update with an edit like:

```diff
  ddl = [
    "",
+    "",
  ]
```

**Review by SarahFrench — COMMENTED** [review]
I had a go manually testing the code changes to help with refreshing my memory about this PR and I realised that the pre_update code isn't making it into the downstream. If the `mmv1/templates/terraform/pre_update/spanner_databse.go.erb` file is renamed to `.../spanner_database.go.erb` I'd expect it to be templated in, but that doesn't seem to happen.

I'm afraid I couldn't work out why that was the case though - maybe due to the update function being a partial update?

**Review by SarahFrench — APPROVED** [review]
Thanks for debugging the issue with the custom code - works as expected now. I was able to provision a database where `ddl = [""]` and then update it to `ddl = ["", ""]`.

I think this resource is a good example of why mixing imperative and declarative ways of provisioning isn't ideal, and I helped contribute to that problem in the past - thank you for navigating it in this PR 🙏

**Pull request merged**